### ISSUE TITLE
keepassxc unmounts google drive folder, then mounts it read-only

### ISSUE THREAD

**Issue by blasco51** [keepassxreboot/keepassxc#4856]
## Overview
Ubuntu 20.04
KeePassXC umounts Google Drive folder while trying to save converted database file on it.
Then opens read-only the database file stored on Google Drive folder.

I'm trying to move to keepassxc from keepass 1
My database is stored on Google Drive. I access it using the Ubuntu standard tool which utilizes Nautilus file manager. Thus Google Drive is seen as a mounted network disk.

## Steps to Reproduce
[NOTE]: # ( Provide a simple set of steps to reproduce this bug. )
1. Database file produced by Keepasdroid on Google Drive
2. Google Drive mounted by Ubuntu 20.04 Nautilus file manager as a network disk
3. KeePassXC  opens database file as a version 1 file to import
4. When saving database on the same Google Drive folder the network disk is unmounted
5. The database file remains with temporary extension and is unusable
6. Re-mount Google Drive
7. Open database file, convert and save it on a local folder
8. Move the .kdbx file to the Google Drive folder
9. Open it, do some mods and close
10. Re-open it, the mods are lost.

## Expected Behavior
Having the database file stored on a local folder or on a mounted network disk should not be different for KeePassXC 

## Actual Behavior
The Google Drive network disk is not properly managed 

## Context
[NOTE]: # ( Give us any additional information you may have. )

[NOTE]: # ( Paste debug info from Help → About here )
KeePassXC - Version 2.4.3
Revision: 5d6ef0c

[NOTE]: # ( Pick choices based on your environment )
Operating System: Linux Ubuntu 20.04
Desktop Env: Gnome
Windowing System: X11

**Comment by phoerious** [comment]
We are not mounting or unmounting anything.

**Comment by blasco51** [comment]
Well maybe Nautilus unmounts Google Drive as a result of improper action by keepassxc

**Comment by blasco51** [comment]
I realized version 2.4.3 from the ubuntu repository is not the latest. I installed version 2.5.4 from ppa:phoerious
The behaviour remains exactly the same :-(

**Comment by droidmonkey** [comment]
The opening of a file as read only is fixed in 2.6.0. Try our beta build.

**Comment by blasco51** [comment]
Well now the database file is saved on google drive but the original file name is lost. It is stored with something that looks like a temporary name: 1BdtRu8yCTCMjJW1O4knX74O0L0mUMeek

**Comment by droidmonkey** [comment]
Ummm that's unique!

**Comment by blasco51** [comment]
I tried that: 
1- open the "strange name" file
2- save database as "new file name" on a local folder
3- open "new file name" 
4- save database as same "new file name" on google drive
Saving fails and google drive is unmounted :-(

**Comment by droidmonkey** [comment]
Do you have safe file saves enabled?

**Comment by blasco51** [comment]
With "safe file saves" enabled I get the unmount problem when saving a new file. Without it I get files saved with the strange temporary name.
I don't have troubles with other apps. I.e. I noticed that when libreoffice edits a google drive file a temporary file name is shown on its window top (that does not happen with a local file). Nonetheless then the file is saved correctly with its name on google drive.

**Comment by droidmonkey** [comment]
So I just tried replicating this on Ubuntu 20.04, using standard Google account features in Gnome. When I tried saving a database to the GVFS mount, GVFS crashed. This is probably why you are seeing it unmount/mount again. It crashed with a null pointer error. This is clearly a bug in GVFS when using QSaveFile type operations. QSaveFile works by writing to a temporary file next to the real file, deleting the real file, then renaming the temporary file as the real file name.

**Comment by droidmonkey** [comment]
Here are the results of my testing:

Safe Saves ENABLED:
* Temporary file is created and written to in the GVFS folder (Passwords.kdbx.XXXXX)
* GVFS crashes
* Temporary file remains indicating that QSaveFile::commit() could not complete

Safe Saves DISABLED:
* Writing a new database to the GVFS folder worked flawlessly

**Comment by blasco51** [comment]
Thank you very much droidmonkey! You pointed very well the problem. But I wouldn't be so sure about a bug in GVFS. Just because other apps that open and save files to GVFS mount point don't have the same problem. You can try and see.

**Comment by droidmonkey** [comment]
Under no circumstances should it be possible to crash GVFS. That is a bug.

**Comment by blasco51** [comment]
Well, I see your point, I will try to open a bug under gitlab.gnome.org
But the "saving an existing database" problems still stands

**Comment by droidmonkey** [comment]
The problem is that Qt receives the file name as something like `/run/user/1000/gvfs/google-drive:host=.../1BdtRu8yCTCMjJW1O4knX74O0L0mUMeek` instead of `google-drive://passwords.kdbx`.

**Comment by blasco51** [comment]
That crazy long name is just the temporary file name. KeePassXC should not use that name but save the file with its original file name. I guess for example Libreoffice does it that way since it shows the crazy long name while editing but uses the original name when saving. You can test and see

**Comment by droidmonkey** [comment]
I have no idea how to get the original file name, it does not appear when choosing the file at all.

**Comment by blasco51** [comment]
Can't you try to avoid deleting the original file? And close by saving the temporary file onto the original?

**Comment by blasco51** [comment]
It looks like I have to give up using KeePassXC because of its incompatibility with Google Drive on Linux :-(

**Comment by Hedroom** [comment]
Hello,

I am having exactly the same issue as blasco51, using what appears to be the same setup (details below) of Ubuntu 20.04 and KPXC 2.6.0.

Using 2.4.3 under Ubuntu 19.04, this was never a problem, and safe saving worked, preserving the filename. Is the above thread an indication that this is purely a GVFS issue in Ubuntu 20.04?


Steps to reproduce:
1. Open a password database on Google Drive
2. Make any changes and save the database
3. Fails to save
4. Disable safe saves
5. Save successfully, but with new random filename

Expected Behavior
Saving a database should preserve the original filename.

Actual Behavior
The filename is changed after every save.

Context
Fresh installation of Ubuntu 20.04
KeePassXC 2.6.0 installed from https://launchpad.net/~phoerious/+archive/ubuntu/keepassxc (repo was added to my sources to keep updating)
Syncing to Google Drive via native Ubuntu Online Accounts service, GVFS via Nautilus file manager

[NOTE]:
KeePassXC - Version 2.6.0
Revision: 0765954

Qt 5.12.8
Debugging mode is disabled.

Operating system: Ubuntu 20.04 LTS
CPU architecture: x86_64
Kernel: linux 5.4.0-40-generic

Enabled extensions:
- Auto-Type
- Browser Integration
- SSH Agent
- KeeShare (signed and unsigned sharing)
- YubiKey
- Secret Service Integration

Cryptographic libraries:
 libgcrypt 1.8.5

[NOTE]:
Operating System: Linux/Ubuntu 20.04
Desktop Env: Gnome
Windowing System: Xorg

**Comment by droidmonkey** [comment]
I'm surprised this worked for you at all in 2.4.3 because it should have opened the database read only.

**Comment by mdeguzis** [comment]
This also is a huge problem for me and how I accidentally lost my main passwords file (now I use a crontab and "drive" for linux to pull them hourly). This is the only app I this issue with. Gnome + System 76 Pop!_OS.

**Comment by Hedroom** [comment]
> I'm surprised this worked for you at all in 2.4.3 because it should have opened the database read only.

I'm very sorry, my mistake - I think it might have been 2.4.1 that saved correctly. I neglected to make a note of the version I was using more successfully before I upgraded my OS and then installed 2.6.0.

**Comment by sashadereh** [comment]
Just adding +1, I also encounter this problem.

**Comment by anuneo** [comment]
Hey, I have this problem as well. Any idea, when it will be fixed? Thank you for your help :)

**Comment by droidmonkey** [comment]
This will be fixed in 2.7.0.

**Comment by Hedroom** [comment]
Thank you, and Merry Christmas!

**Comment by romainjacquet** [comment]
I report that I had a very similar problem with Keepass 2.4.3. When opening keepass database stored on a Google Drive network share, the database is open read-only. In the title of the tab, I got "MyDATABASE [Read-only]".
I try the original application Keepass and the problem is really worse. When saving the file, nothing is saved and the file is truncated to zero. 

Keepass is using Mono runtine for IO operations and KeepassXC uses the native functions provided by QtCore. Because both Keepass and Keepassxc are concerned, I think so the problem could come from the underlying VFS. I mean the code of gvfsd-google daemon, which is part of gvfs-backends 1.44. 
I'm going to perform more tests.

**Comment by janlucaklees** [comment]
Having the same issue with Gnome on Manjaro Linux. I think it's the same as described here. All I can add is that it seems KeePassXC, or something in between KeePassXC and Google Drive, tries to save the file not under the original name, but under the file id that Google Drive uses internally.

Thus a new file is created which of course gets a new id. So that opening and saving it again, will result in yet another file created.

I mean, knowing that, one could just name their password db just as the Google Drive file id.

**Comment by droidmonkey** [comment]
The crazy thing about this bug is that the file chooser gives us the path to the file ID version. That is a backend implementation detail of gvfs that we shouldn't have to deal with. I've tried several methods to get the front-end path to the file using Qt and failed to get it to work. Hence why this is still open.

**Comment by janlucaklees** [comment]
But they can show the files names correctly in their file explorer. There's gotta be a way. :thinking:

**Comment by droidmonkey** [comment]
In file explorer you are seeing the "front end" of GVFS which is user focused. The file chooser dialog, however, returns the "back end" file path for GVFS which should really not be exposed to programs. Everything should go through the front door, but it doesn't, so any program that does not write directly to the provided file node in the back door will have this error. The way saving works in KeePassXC is we write a temporary file right next to the real one, delete the real one, and rename the temporary file. This prevents data loss due to interrupted file writes. This also prevents us from using the GVFS backdoor and why your files get named the file ID instead.

**Comment by janlucaklees** [comment]
Yeah, this is what I just found out over at the GNOME GitLab project.

So basically, besides customizing KeePassXC to work with GVfs there's nothing we can do about it.

**Comment by droidmonkey** [comment]
Bingo

**Comment by spoonerandrew** [comment]
Just to add that the same behaviour is happening on Fedora 34 with KeePassXC 2.6.4

**Comment by rattkin** [comment]
is there an alternative client that works better with google drive? Or some workaround?

**Comment by robellegate** [comment]
As @droidmonkey has already stated, the weird file name is the [back-end file name used by Google Drive][1]. The way Gnome applications such as Nautilus gets the file name is from the `standard::display-name` metadata. You can [see this information using `gio`][2]:

    gio list -a "standard::display-name" 'google-drive://<your_email>@gmail.com/My Drive'

Perhaps the display name could be used in place of the file path in the UI? I'm currently digging through the gvfs source code to see if this metadata can be exposed another way.
[1]: https://gitlab.gnome.org/GNOME/gvfs/-/issues/402#note_515990
[2]: https://gitlab.gnome.org/GNOME/gvfs/-/issues/402#note_668213

**Comment by droidmonkey** [comment]
For information, I added a fix for this in the develop branch. You can choose to "direct write" the database file which is the only way to save files correctly using GVFS.

**Comment by lawrence-laz** [comment]
> For information, I added a fix for this in the develop branch. You can choose to "direct write" the database file which is the only way to save files correctly using GVFS.

How dangerous is this? If connection drops during save, is .kdbx file screwed?

**Comment by droidmonkey** [comment]
Yes, very dangerous from that perspective. But Google would have a backup of your database server side if that happened.

**Comment by lawrence-laz** [comment]
Oh, as long as the file can be recovered it's fine then I guess

**Comment by michaelk83** [comment]
> Oh, as long as the file can be recovered it's fine then I guess

You should always keep your own backups as well, regardless of which save method you use.

**Comment by droidmonkey** [comment]
Fixed with #6594

**Comment by Excrubulent** [comment]
I just had this same issue on KeepassXC 2.7.1 on Ubuntu 22.04 using the built-in Gnome integration of Google Drive.

Let me know if you want any more details.

EDIT:  I should also mention the UI for KeepassXC hangs when attempting to open the file on google drive from within the app, using the Open Database option.  It opens properly when opening it from the file manager using the Open With feature.  Not sure if this is a clue to what's going on but it might be relevant.

**Comment by campey** [comment]
I just had this issue on KeepassXC 2.6.6 [default] on Ubuntu 22.10, using Gnome google drive like @Excrubulent above.

It ended up renaming, and deleting the file. (In a way that permanently deleted the file on Google drive, and Google wwon't take service requests to restore such permanently deleted app backup files it seems) It was rather stressful as I hadn't been diligent with backups for a while. Luckily my phone still had a local cache and I was able to export a backup from there.

I have installed 2.7.4 and will retry and share experiences.

Please ask if you'd like any specific info, or if there is a work-around.

**Comment by campey** [comment]
Okay, so from the new update: 2.7.4
I observed with safe saves enabled, that a new file appeared (long google id), then the old one disappeared, and then an error about failing to write the database appeared. Then it prompted me to enable unsafe saves which I did and the file was saved named with the long google id.

This feels similar to what @droidmonkey described in 2020 https://github.com/keepassxreboot/keepassxc/issues/4856#issuecomment-643780553

**Comment by campey** [comment]
![image](https://user-images.githubusercontent.com/1193860/204291364-a103b7a1-136d-4e60-88b9-4bcb0f82b215.png)

[EDIT] In trying to upgrade to 2.7.4 installed 2 different keepassxc's side-by-side.
2.6.6 installed from source: "ubuntu-kinetic-universe" installed as "KeePassXC"
2.7.4 installed form source: "snap store" installed as "keepassxc"

2.7.4 would not even open my file (hung after selecting the file) so, unfortunately, can't test this.
2.6.6 saving with "safely save" unchecked results in the file disappearing, the drive folder activity shows only that the file was "permanently deleted", with safe saves, get the temp file and rename issues exactly as reported by @droidmonkey, so as would be expected I suppose.

"work-around": I will continue using KeepassXC 2.6.6 on ubuntu to access passwords, but not save any new passwords in via there.

**Comment by droidmonkey** [comment]
Use our flatpak deployment or appimage, not snap. Snaps are garbage. The 'better' alternative saving strategy was introduced in 2.7.0, so 2.6.6 is considered buggy at best.

**Comment by campey** [comment]
fwiw I just tried with appimage. It uses a different file browser that didn't have a way to browse to the google drive (no links to the "gnome" google drive), and the open database dialogue keeps going unresponsive.

When I pasted in the google drive link from "Files" browser on ubuntu, /run/user/1000/gvfs/google-drive:host=[domain],user=campey/[long_id] then I see the files with their google files ids in the browser there, and it hangs again.

**Comment by droidmonkey** [comment]
Welcome to the state of the linux desktop... ugh

**Comment by campey** [comment]
For completeness, I tried with flatpak, selecting the database file caused it to hang (left for a minute or so) without asking for password or key file.
Then based on another comment I read, I tried to open it using the "open with..." context menu, which got me to the password screen, and I could successfully choose key file and enter password.
But then, it didn't open the file, failing with the message: "Error while reading the database: No root group"

So, I'm going back to 2.6.6 for at least being able to read, please feel free to let me know if you'd like any more tests/info.

**Comment by droidmonkey** [comment]
That error message indicates a corrupt database file or truncated reading

**Comment by campey** [comment]
Further investigation: I tried in 2.7.4 (flatpak and ppa:phoerious/keepassxc) and kept getting the "no root group" message, copied to a local file to see if a GVFS error there, still persisted.

Removed the phoerious repository, reinstalled 2.6.6 from ubuntu kinetic, and can successfully read the file from both gvfs and local file. Hope this information is in some way useful, but I think I'm done playing for now :) ... also realise this probably needs another issue, they were just encountered in trying to test the current issue.

Please do ask if there is any specific debugging you would like.

**Comment by droidmonkey** [comment]
I recommend you make a change to the database in 2.6.6, save it locally, and then open in 2.7.4. There are no code changes between 2.6.6 and 2.7.4 that would relate to opening a database where the root group is concerned. Very curious how this is even possible.

**Comment by rafabr4** [comment]
Hello. I'm using Pop!_OS 22.04 LTS with KeePassXC 2.7.4 installed from flatpak. I'm using the integrated online accounts feature for syncing my Google Drive.

When trying to open a database from the online location, KeePassXC simply hangs: the UI doesn't change, it doesn't show the prompt for the password, and it becomes totally unresponsive (need to kill the process). When trying to open the same database file from a local folder, it immediately shows the password prompt and loads it successfully.

I've tried both enabling and disabling the "Use alternative saving method".

Can you think of any possible reasons? I'm willing to do testing/logging if needed.

Edit: I posted it here because I was looking for a solution to the random file renaming of the database.

**Comment by droidmonkey** [comment]
This sounds like a GVFS or flatpak bridge to GVFS problem. Does this happen to other flatpak apps accessing files on GVFS?

**Comment by TheDogHusky** [comment]
Using 2.7.9, I cannot open a keepass2 database through GVFS on nautilus. 
https://safe.classydev.fr/2024_07_19_18_24_47-u8kIQrjkYL0ikMn6sxV2nYByppSQBrjG.mkv here is a video of the issue.
Going through /run/user/1000/gvfs/<drive folder> does not seem to be an option for me as it takes forever to load and is just a mess (and as said by other people seems to be "temporary" names/folders that may change so f*ck up my path)

**Comment by LinuxfarmerHH** [comment]
I am on KeepassXC 2.6.6 on Linuxmint 21.3 and Zorin-OS 7.1 (Ubuntu 22.04lts based)

Same as described before, saving to gdrive or my local NAS (AFP or SMB protocol) will mangle the database filename to something crude without .kdbx suffix. This renders the usability down, you have to save on your local disk and copy the .kdbx by hand to a share and reverse on the other computer. Okay it's better than sending a fax, but it's not cool.

GVFS Info
ii  gvfs:amd64      1.48.2-0ubuntu1 amd64        userspace virtual filesystem ->
ii  gvfs-backends   1.48.2-0ubuntu1 amd64        userspace virtual filesystem ->
ii  gvfs-common     1.48.2-0ubuntu1 all          userspace virtual filesystem ->
ii  gvfs-daemons    1.48.2-0ubuntu1 amd64        userspace virtual filesystem ->
ii  gvfs-fuse       1.48.2-0ubuntu1 amd64        userspace virtual filesystem ->
ii  gvfs-libs:amd64 1.48.2-0ubuntu1 amd64        userspace virtual filesystem -

**Comment by droidmonkey** [comment]
Take it up with gvfs please

**Comment by Josh68** [comment]
> I am on KeepassXC 2.6.6 on Linuxmint 21.3 and Zorin-OS 7.1 (Ubuntu 22.04lts based)
> 
> Same as described before, saving to gdrive or my local NAS (AFP or SMB protocol) will mangle the database filename to something crude without .kdbx suffix. This renders the usability down, you have to save on your local disk and copy the .kdbx by hand to a share and reverse on the other computer. Okay it's better than sending a fax, but it's not cool.
> 
> GVFS Info ii gvfs:amd64 1.48.2-0ubuntu1 amd64 userspace virtual filesystem -> ii gvfs-backends 1.48.2-0ubuntu1 amd64 userspace virtual filesystem -> ii gvfs-common 1.48.2-0ubuntu1 all userspace virtual filesystem -> ii gvfs-daemons 1.48.2-0ubuntu1 amd64 userspace virtual filesystem -> ii gvfs-fuse 1.48.2-0ubuntu1 amd64 userspace virtual filesystem -> ii gvfs-libs:amd64 1.48.2-0ubuntu1 amd64 userspace virtual filesystem -

Same issue running Manjaro with Cinnamon DE and using Online Accounts to link my Google account, including Drive. On KDE, seems to work fine, so sounds like gvfs is the issue.

**Comment by appovo** [comment]
For those landing here - checking this box did the trick for me in 2.7.9

![Image](https://github.com/user-attachments/assets/4a48971a-7ccc-45fd-be1c-f148736d1052)

**Comment by JanRou** [comment]
> For those landing here - checking this box did the trick for me in 2.7.9
> 
The file IS NOT UPDATED on the Google drive for me.

It seems not to solve the problem that KeePassXC can't update the database on Google Drive attached the filesystem with Gnome.

It moves my password database from Google drive into a file under a strange folder /run/user/1000/doc/73a65601/0AMBlzg1AS2U9Uk9PVA. Now I'm stuck, because I don't how to instruct KeePassXC to write the changed file to my Google drive with the original name. I want the file updated with my changes.

**Comment by droidmonkey** [comment]
You need to enable direct write saves in the application settings

**Comment by JanRou** [comment]
Ohh I'm new to linux, so please show me how to change to direct application settings.

**Comment by JanRou** [comment]
I have played with the settings and concluded none works safely. I got the must-not-happen scenario: I lost the database. Fortunately I had made a copy before playing with KeePassXC.

Conclusion:  Don't trust KeePassXC in combination with a google drive mounted with Gnome. It's view only from Linux. KeePassXC works great with local database.

**Comment by droidmonkey** [comment]
You don't have direct write saves enabled.

**Comment by saru-man** [comment]
I had that option enabled and my KeePassXC 2.7.10 behaves exactly the same as @JanRou's:

1. "Alternate Saving Option" disabled: KeePassXC creates a new file called something like 0AMBlzg1AS2U9Uk9PVA, but removes the old database file
2. "Alternate Saving Option" enabled with "Temporary file saves":  KeepAassXC creates a new file called something like qt_temp.plKyDW
3.  "Alternate Saving Option" enabled with "Direct-write saves": KeePassXC's not able to save and throws error messages on and on so I had to kill the KeePassXC instance - resulting in a 0 byte sized database file and some new qt_temp.* files

Edit:

Played around a bit and found this explanation explaining the way how files are stored inside of Google Drive and what that means for GVFS interacting with Google Drive: https://discussion.fedoraproject.org/t/address-to-google-drive-via-terminal/75841 - especially https://discussion.fedoraproject.org/t/address-to-google-drive-via-terminal/75841/5

That inspired me to edit ~/.cache/keepassxc/keepassxc.ini and tried if files are accessible even when using the display name from Google Drive (sorry, directory names are in german but I think you get the idea):

`LastActiveDatabase="/run/user/1000/gvfs/google-drive:host=gmail.com,user=XXXX/Meine Ablage/KeePass/KeePass.kdbx"
LastDatabases="/run/user/1000/gvfs/google-drive:host=gmail.com,user=XXXX/Meine Ablage/KeePass/KeePass.kdbx"
LastOpenedDatabases="/run/user/1000/gvfs/google-drive:host=gmail.com,user=XXXX/Meine Ablage/KeePass/KeePass.kdbx"`

And here we are! KeePassXC opens the database file correctly and is still able to save a backup name with it's correct file name. With the settings mentioned by @droidmonkey  saving the database file is possible, too but it's real sloooooow so it can be used more as a tech demo than in production state. But that's a GVFS related issue and has nothing to do with KeePassXC itself.

**Comment by droidmonkey** [comment]
I agree that GVFS is ridiculously slow, I just tested usage in the latest GNOME. That is definitely "their problem" though and not ours. They really should be using a local directory that is watched by a daemon rather than a direct-mount option. The current way of interfacing with files is ridiculous.
